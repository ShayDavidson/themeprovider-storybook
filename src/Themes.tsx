import { List } from "immutable";
import * as React from "react";
import { Button, Row } from "./components/Button";
import { Theme } from "./types/Theme";
import SvgIcon from "./components/SvgIcon";

export interface IThemeProps {
  channel: any;
  api: any;
  active: boolean;
}

interface IButtonProps {
  onSelectTheme: (theme: Theme) => void;
  theme: Theme;
  onOpenModal: () => void;
  themes: List<Theme>;
}

const BaseComponent: React.FunctionComponent<IButtonProps> = ({
  onSelectTheme,
  themes,
  theme,
  onOpenModal,
}) => (
  /* tslint:disable:jsx-no-multiline-js jsx-no-lambda*/
  <Row>
    {themes
      .map((th, i) => (
        <Button
          isSelected={th === theme}
          key={i}
          onClick={() => onSelectTheme(th)}
        >
          <span>{th.name}</span>
          <SvgIcon style={{ marginLeft: '1em' }} name="info" onClick={() => onOpenModal()} />
        </Button>
      ))
      .toArray()}
  </Row>
);

export const Themes: React.FunctionComponent<IThemeProps> = ({
  channel,
  api,
  active,
}) => {
  const [theme, setTheme] = React.useState(null);
  const [themes, setThemes] = React.useState(List());

  const onReceiveThemes = (newThemes: Theme[]) => {
    // tslint:disable-next-line: no-shadowed-variable
    const themes = List(newThemes);
    const themeName = api.getQueryParam("theme");
    setThemes(themes);
    if (themes.count() > 0) {
      // tslint:disable-next-line: no-shadowed-variable
      const theme =
        themes.find((t) => t.name === themeName) || themes.first();
      setTheme(theme);
      channel.emit("selectTheme", theme);
    }
  };

  const onSelectTheme = (theme: Theme) => {
    setTheme(theme);
    api.setQueryParams({ theme: theme.name });
    channel.emit("selectTheme", theme);
  };

  const onOpenModal = () => {
    channel.emit("openModal", true);
  }

  React.useEffect(() => {
    channel.on("setThemes", onReceiveThemes);
    return () => {
      channel.removeListener("setThemes", onReceiveThemes);
    };
  }, []);

  return theme && active ? (
    <BaseComponent
      onOpenModal={onOpenModal}
      onSelectTheme={onSelectTheme}
      themes={themes}
      theme={theme}
    />
  ) : null;
};
