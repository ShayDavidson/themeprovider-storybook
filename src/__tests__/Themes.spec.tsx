import { mount } from 'enzyme';
import * as React from "react";
import { spy } from "sinon";
import { Themes } from "../Themes";

describe("Themes spec", () => {
  it("Should render proper", () => {
    const channel = {
      emit: spy(),
      on: spy(),
      removeListener: spy(),
    };

    const component = mount(
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      <Themes channel={channel} active={true} theme={{ someTheme: {} }} />,
    );

    expect(component.debug()).toMatchSnapshot();
    expect(channel.on.called).toBeTruthy();
    expect(channel.emit.notCalled).toBeTruthy();

    component.unmount();
    expect(channel.removeListener.called).toBeTruthy();
  });
});
