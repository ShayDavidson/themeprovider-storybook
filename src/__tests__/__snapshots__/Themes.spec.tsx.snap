// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themes spec Should render proper 1`] = `"<Themes channel={{...}} active={true} theme={{...}} />"`;
